### DISCUSSION TITLE
Questions about fields

### DISCUSSION THREAD

**Discussion by tim-coder** [cachapa/tudo#39] [General]
Thank you so much for creating the tudo example. I'm learning a lot from it.

I want to make sure I understand your implementation of CRDT. Please let me know if this correct:

To use CRDT for a generic table, the following fields are added:

 - is_deleted:  used for soft deletes
 - node_id: contains the ID of the node creating/modifying the record.
 - hlc:  the timestamp for record creation/modification appended with the ID of the mutating node
 - modified: the timestamp for the record entry/update on the current node (not changes to the data), appended with the node ID where the record is being saved

If this is correct, then I have two questions:
1. Isn't appending the node ID to the modified field redundant since the node writing the record would know its own ID?
2. Isn't appending the node ID to the hlc entry redundant, since that value will always be equal to the value of the node_id field? Or is it needed for compatibility with earlier versions of the code?

**Comment by cachapa** [comment]
> Isn't appending the node ID to the modified field redundant since the node writing the record would know its own ID?

Yes, it's done for performance since it allows searching the database using direct, rather than partial string comparisons.

> Isn't appending the node ID to the hlc entry redundant, since that value will always be equal to the value of the node_id field? Or is it needed for compatibility with earlier versions of the code?

If you mean in the `hlc` column, then it's not always the same - that node id will be the one which updated the record last.

If you mean the `modified` column, you're right, it will always be the local node id, and it's theoretically redundant. The reason it stayed is twofold: 1) it just uses a standard HLC rather than needing a variation of it with an empty node id; 2) it is a convenient way to implicitly store the local node id directly with the data: the canonical time is simply the largest `modified` hlc, and the local node id is taken directly from canonical time.

**Comment by tim-coder** [comment]
Yes, that's what I meant - the `modified` column.

Thanks so much for the info.
